---
title: "scTSS Tutorial (Paired-end)"
output:
  md_document:
    variant: markdown_github
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

<!-- # Overview of scTSS -->

<!-- scTSS is a package that can leverage the 5' single-cell RNA-seq data (both paired-end and single-end data) to study the transcription start site (TSS) at single-cell resolution. scTSS has two major steps, 1) TSS prediction and quantification; 2) differential TSS usage (DU) analysis. Our pipeline considers a general scenario where multiple single-cell samples (e.g., each sample is from a different donor or biological condition) are available, and a joint analysis is to be performed.  -->

<!-- To download scTSS from github and load package, please follow: -->

<!-- ```{r eval=FALSE} -->
<!-- devtools::install_github("https://github.com/ShweiSTAT/scTSS") -->
<!-- library(scTSS) -->
<!-- ``` -->

<!-- Since predicting TSS clusters based on paired-end and single-end data are different, this tutorial is split into two parts based on the different types of data used.  -->

<!-- The data used in this tutorial can be downloaded [here](https://www.dropbox.com/sh/prmvy6l4e8wh390/AACfyKto6U_Zh7gZ53bFYP4ka?st=up3bu8ud&dl=0). -->


# Preprocessing

The paired-end 5' single-cell RNA-seq data contains both read 1 and read 2. The read 1 contains the direct information of the transcription start site (TSS). Most of the TSS analysis tools are capable of predicting TSS clusters based on the mapped genomic positions of the 5' of the read 1. Thus, scTSS will utilize the existing tools to predict TSS clusters for each sample.  

![Example Image](https://www.dropbox.com/scl/fi/chiqbkljcrw7czsk0nhtv/pairedEndDemo.pdf?rlkey=m9h5oq2y48wxwiqbf1q0u1um2&st=sb42z4qa&dl=0)

For the paired-end data, the preprocessing starts with the raw reads (**.fastq** files) from 5' single-cell RNA-seq experiments.

 + First, the reads should be aligned to reference genome by an alignment tool for single-cell RNA sequencing data. We assume this alignment is performed sample by sample. For example, for the reads generated by 10x Genomics 5' Gene Expression assay, we use Cell Ranger to perform alignment for each one of the samples. At the end of the alignment step, we obtain the aligned **.bam** file and the corresponding **.bai** file for each sample.

 + Second, we need to use the **.bam** files to obtain predictions for the TSS clusters by a bioinformatics tool compatible with on-site sequencing data. We recommend using Homer findPeaks. Please refer to this [link](http://homer.ucsd.edu/homer/ngs/peaks.html) for details. After running findPeaks, we should obtain a **.txt** file that contains the called peaks for each sample. Please remove the lines commented by `#`, except for the header line. For the header line, please just remove the `#` itself and retain other header information. If the file is from other peak calling tools, please make sure the file contains the following four columns and they are separated by ``\t``. Please refer to this [file](https://www.dropbox.com/s/dsxh9ggfas373pr/example_peak.txt?dl=0) for a specific example.
    - PeakID: a distinct ID for each peak
    - chr: the chromosome where the peaks located
    - start: the start location (on genome) of the peak
    - end: the end location (on genome) of the peak

+ Third, we need to prepare the cell barcode information for each sample. The barcodes for each sample should be stored in a separate **.csv** file as a single column without header. Meanwhile, please do not include rownames in any files. Please refer to this [file](https://www.dropbox.com/s/ihfkfu97xwf6m55/example_barcode.csv?dl=0) for a specific example.

+ Last, a **.gtf** for reference genome is needed. This file should match the one used in the read alignment.


<!-- ## TSS prediction  -->

<!-- `TSSpredict` is done sample by sample, and requires four input files: -->

<!--   + The aligned **.bam** file of the sample. Noting that the corresponding **.bai** file should be stored under the same directory. -->
<!--   + The **.gtf** file of the reference genome. -->
<!--   + The **.txt** file for the peak calling results of this sample. -->
<!--   + The **.csv** file storing the cell barcode information.  -->

<!-- Besides the input files, `TSSpredict` also requires the following important parameters: -->

<!--   + `tss_filtering_distance` is an integer and controls the minimum distance between any two predicted TSSs. This distance is used for filtering TSS candidates to account for splicing events. We recommend set this parameter to the length of the reads. The default is set to 150 bp. -->

<!--   + `gene_of_interest` is a vector of characters and specifies the genes for which the TSS prediction should be performed. The default is set to `NULL`, which means all the genes in the **.gtf** file are considered. The input gene names should be compatible to the **.gtf** file.  -->

<!--   + `if_paired` is TRUE/FALSE and specifies if the data is paired-end (TRUE) or single-end (FALSE). The default is TRUE. -->

<!--   + `ncore` is an integer and specifies the number of cores used for parallel computation. The default is 1. -->

<!-- Below we show an example to run `TSSpredict` on 20 genes. This example uses a subset of a reads generated by 10X Genomics 5' single-cell RNA-seq. The data is paired-end and read length is 150 bp. The **.gtf** file used in this example is the GRCh38 reference. -->
<!-- ```{r eval=FALSE} -->
<!-- bam <- "./example_BAM.bam" -->
<!-- gtf <- "./example.gtf" -->
<!-- tss <- "./example_peak.txt" -->
<!-- barcodes <- "./example_barcode.csv" -->
<!-- genes <- c("ENSG00000141956","ENSG00000141959", -->
<!--            "ENSG00000142149","ENSG00000142156", -->
<!--            "ENSG00000142166","ENSG00000142168", -->
<!--            "ENSG00000142173","ENSG00000142178", -->
<!--            "ENSG00000142182","ENSG00000142185", -->
<!--            "ENSG00000142188","ENSG00000142192", -->
<!--            "ENSG00000142197","ENSG00000142207", -->
<!--            "ENSG00000154639","ENSG00000154640", -->
<!--            "ENSG00000154642","ENSG00000154645", -->
<!--            "ENSG00000154646","ENSG00000154654") -->

<!-- # predict TSS -->
<!-- TSS_predict_result <- TSSpredict(tss_filtering_distance = 150, -->
<!--                                  gene_of_interest = genes, -->
<!--                                  if_paired = TRUE, -->
<!--                                  ncore = 1, -->
<!--                                  bam_path = bam, -->
<!--                                  anno_path = gtf, -->
<!--                                  tss_path = tss, -->
<!--                                  barcodes_path = barcodes) -->

<!-- ``` -->

<!-- After running `TSSpredict`, we obtain a TSS count matrix stored as a `data.table` object. Each row represents a predicted TSS, each column represents a single cell, and each element in the matrix represents the read count of a predicted TSS in a cell. The first 6 predicted TSSs for the first 6 cells in one of the samples from the above example is shown below. The predicted TSSs are named in the format of [chromosome]:[gene]:[location]:[strand] (e.g. chr21:ENSG00000142166:33335591:+). -->

<!-- ```{r eval=FALSE} -->
<!-- head(TSS_predict_result[,1:7]) -->
<!-- ``` -->

<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE} -->
<!-- TSS_predict_result <- readRDS("/Users/shiweifu/Desktop/exampleTSSpredict.rds") -->
<!-- head(TSS_predict_result[,1:7]) -->
<!-- ``` -->



<!-- ## Differential TSS usage (DU) test -->

<!-- Before we conduct DU test for multiple samples, we need to perform two preparation steps. 1) TSS merging for TSS count matrices from multiple samples, and 2) converting TSS count to TSS usage. -->

<!-- #### TSS merging -->

<!-- TSS merging is achieved by function `TSSmerge`. The input of `TSSmerge` should be a `list` of TSS count matrices obtained by `TSSpredict`. In the following example, we have 10 TSS count matrices stored in an object named `unmerged_TSS_matrices`. The first 5 matrices are based on B cells from 5 different donors and the last 5 matrices are based on NK cells from the same 5 donors.  -->

<!-- ```{r eval=FALSE} -->
<!-- # read in the list of 10 TSS count matrices -->
<!-- unmerged_TSS_matrices <- readRDS("./TSS_matrices.rds") -->

<!-- # perform TSS merging -->
<!-- mergedTSS_matrices <- TSSmerge(tss_matrices = unmerged_TSS_matrices, -->
<!--                                tss_filtering_distance = 150, -->
<!--                                ncore = 2) -->
<!-- ``` -->

<!-- `TSSmerge` returns a `list` of merged TSS count matrices. In our example, the returned `list` contains 10 merged matrices. Compared with the input, the TSS names are updated to TSS regions in the merged TSS matrices. The name of TSS regions follows this format: [chromosome]:[gene]:[start location]:[end location]:[strand] (e.g. chr10:ENSG00000077150:102394303:102394331:+). The first 4 cells for the first 6 merged TSS regions in one of the samples are shown below. -->

<!-- ```{r eval=FALSE} -->
<!-- head(mergedTSS_matrices$`B:sample1`[,1:5]) -->
<!-- ``` -->

<!-- ```{r echo=FALSE} -->
<!-- mergedTSS_matrices <- readRDS("/Users/shiweifu/Desktop/mergedTSS_matrices.rds") -->
<!-- head(mergedTSS_matrices$`B:sample1`[,1:5]) -->
<!-- ``` -->

<!-- #### Converting TSS counts to usage -->
<!-- After TSS merging, we need to convert the TSS count to TSS usage to perform DU test. This can be achieved by the function `CountToUsage`. The example below converts the TSS count matrices obtained from the above TSS merging step to TSS usage matrices. -->

<!-- ```{r eval=FALSE} -->
<!-- # convert TSS count to TSS usage -->
<!-- TSS_usage_matrices <- list() -->
<!-- for( i in 1:length(mergedTSS_matrices)){ -->
<!--   TSS_usage_matrices[[i]] <- -->
<!--     CountToUsage(tss_count_matrix = mergedTSS_matrices[[i]]) -->
<!--   } -->
<!-- names(TSS_usage_matrices) <- names(mergedTSS_matrices) -->
<!-- ``` -->

<!-- After converting TSS count to TSS usage, we show the usage for the first 6 TSS regions below. -->

<!-- ```{r eval=FALSE} -->
<!-- head(TSS_usage_matrices$`B:sample1`[,1:5]) -->
<!-- ``` -->

<!-- ```{r echo=FALSE} -->
<!-- TSS_usage_matrices <- readRDS("/Users/shiweifu/Desktop/TSS_usage_matrices.rds") -->
<!-- head(TSS_usage_matrices$`B:sample1`[,1:5]) -->
<!-- ``` -->


<!-- #### DU test -->

<!-- After we obtained the TSS usage matrices, we can perform DU test. This can be done by the function `TSSduTest`. This function takes two inputs: -->

<!--   + `DU_input`, a `list` object, each element of this `list` should be a TSS usage matrix obtained from `CountToUsage`.  -->
<!--   + `sample_label`, a vector of characters, specifies the conditions on which the test is performed. The length of this vector should be same as the `DU_input` and their order should also match. -->

<!-- For example, the code below compares the TSS usage between the B cells and NK cells from the 5 donors. -->
<!-- ```{r eval=FALSE} -->
<!-- du_rst <- TSSduTest(DU_input = TSS_usage_matrices, -->
<!--                     sample_label = c(rep("B",5), rep("NK",5)), -->
<!--                     ncore = 1) -->
<!-- ``` -->

<!-- The `TSSduTest` outputs a `data.frame` containing the *p*-values and other statistics. -->

<!-- + "fold_change" provides the fold change of mean TSS usage between the conditions that have the highest and lowest mean TSS usage. -->

<!-- + "max_label" provides the condition that has the maximum average TSS usage. -->

<!-- + "max_meanUsage" provides the average TSS usage for the condition that has the maximum average TSS usage. -->

<!-- + "min_label" provides the condition that has the minimum average TSS usage. -->

<!-- + "min_meanUsage" provides the average TSS usage for the condition that has the minimum average TSS usage. -->

<!-- + "num_of_cells" provides the number of cells where this TSS region is activated for use. -->

<!-- A part of the DU result is shown below. -->
<!-- ```{r eval=FALSE} -->
<!-- DU_rst$DU_result[1:3,] -->
<!-- ``` -->

<!-- ```{r echo=FALSE} -->
<!-- DU_rst <- readRDS("/Users/shiweifu/Desktop/DU_rst.rds") -->
<!-- DU_rst$DU_result[1:3,] -->
<!-- ``` -->
